machine_learning/loss_functions.py

```diff
@@ -261,6 +261,49 @@ def mean_squared_error(y_true: np.ndarray, y_pred: np.ndarray) -> float:
     return np.mean(squared_errors)
 
 
+def mean_absolute_error(y_true: np.ndarray, y_pred: np.ndarray) -> float:
+    """
+    Calculates the Mean Absolute Error (MAE) between ground truth (observed)
+        and predicted values.
+
+    MAE measures the absolute difference between true values and predicted values.
+
+    Equation:
+    MAE = (1/n) * Σ(abs(y_true - y_pred))
+
+    Reference: https://en.wikipedia.org/wiki/Mean_absolute_error
+
+    Parameters:
+    - y_true: The true values (ground truth)
+    - y_pred: The predicted values
+
+    >>> true_values = np.array([1.0, 2.0, 3.0, 4.0, 5.0])
+    >>> predicted_values = np.array([0.8, 2.1, 2.9, 4.2, 5.2])
+    >>> np.isclose(mean_absolute_error(true_values, predicted_values), 0.16)
+    True
+    >>> true_values = np.array([1.0, 2.0, 3.0, 4.0, 5.0])
+    >>> predicted_values = np.array([0.8, 2.1, 2.9, 4.2, 5.2])
+    >>> np.isclose(mean_absolute_error(true_values, predicted_values), 2.16)
+    False
+    >>> true_labels = np.array([1.0, 2.0, 3.0, 4.0, 5.0])
+    >>> predicted_probs = np.array([0.3, 0.8, 0.9, 5.2])
+    >>> mean_absolute_error(true_labels, predicted_probs)
+    Traceback (most recent call last):
+    ...
+    ValueError: Input arrays must have the same length.
+    """
+    if len(y_true) != len(y_pred):
+        raise ValueError("Input arrays must have the same length.")
+
+    if isinstance(y_true, np.ndarray) and isinstance(y_pred, np.ndarray):
+        return np.mean(abs(y_true - y_pred))
+    else:
+        try:
+            return np.mean(abs(np.asarray(y_true) - np.asarray(y_pred)))
+        except ValueError as error:
+            raise ValueError("Could not convert input to NumPy array.") from error
```

Comment:
```suggestion
    return np.mean(abs(y_true - y_pred))
```

Personally I feel that it's OK to assume that inputs will be numpy arrays, so this check is unnecessary IMO. While real-world software would need to handle such invalid inputs, this repo is primarily for educational purposes, and we want to balance clarity with functionality.